src/query_builder.rs

```diff
@@ -0,0 +1,737 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![macro_use]
+use std::collections::{
+    BTreeMap,
+};
+use std::rc::Rc;
+
+use indexmap::{
+    IndexMap
+};
+
+use indexmap;
+
+use mentat_core::{
+    Entid,
+    NamespacedKeyword,
+    TypedValue,
+    TypedValueConverter,
+    ValueType,
+};
+
+use ::{
+    DateTime,
+    Queryable,
+    QueryInputs,
+    QueryOutput,
+    Store,
+    Uuid,
+    Utc,
+    Variable,
+};
+
+use errors::{
+    Result,
+};
+
+pub struct QueryBuilder<'a> {
+    sql: String,
+    values: BTreeMap<Variable, TypedValue>,
+    types: BTreeMap<Variable, ValueType>,
+    store: &'a mut Store,
+}
+
+impl<'a> QueryBuilder<'a> {
+    pub fn query<T>(store: &'a mut Store, sql: T) -> Result<QueryBuilder> where T: Into<String> {
+        Ok(QueryBuilder { sql: sql.into(), values: BTreeMap::new(), types: BTreeMap::new(), store })
+    }
+
+    pub fn bind_value<T>(mut self, var: &str, value: T) -> Self where T: Into<TypedValue> {
+        self.values.insert(Variable::from_valid_name(var), value.into());
+        self
+    }
+
+    pub fn bind_ref_from_kw(mut self, var: &str, value: NamespacedKeyword) -> Result<Self> {
+        let entid = self.store.conn().current_schema().get_entid(&value).ok_or(ErrorKind::UnknownAttribute(value.to_string()))?;
```

Comment:
I'm a bit leery of having to fetch the current schema for each attempt to bind; not only is it not free, but the schema might change between bindings!

Is there a reason why you wouldn't buffer these and resolve them in `execute`?